drivers/can/can_mcux_flexcan_fd.c

```diff
@@ -0,0 +1,1091 @@
+		data->timing_data.sjw = sjw_data_backup;
+	}
+
+	return 0;
+}
+
+static int mcux_flexcan_fd_get_capabilities(const struct device *dev, can_mode_t *cap)
+{
+	ARG_UNUSED(dev);
+
+	*cap = CAN_MODE_NORMAL | CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY | CAN_MODE_3_SAMPLES |
+	       CAN_MODE_FD;
+
+	return 0;
+}
+
+static int mcux_flexcan_fd_start(const struct device *dev)
+{
+	const struct mcux_flexcan_fd_config *config = dev->config;
+	struct mcux_flexcan_fd_data *data = dev->data;
+	flexcan_timing_config_t timing;
+	int err;
+
+	if (data->started) {
+		return -EALREADY;
+	}
+
+	if (config->phy != NULL) {
+		err = can_transceiver_enable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to enable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	/* Clear error counters */
+	config->base->ECR &= ~(CAN_ECR_TXERRCNT_MASK | CAN_ECR_RXERRCNT_MASK);
+
+	/* Delay this until start since setting the timing automatically exits freeze mode */
+	timing.preDivider = data->timing.prescaler - 1U;
+	timing.rJumpwidth = data->timing.sjw - 1U;
+	timing.phaseSeg1 = data->timing.phase_seg1 - 1U;
+	timing.phaseSeg2 = data->timing.phase_seg2 - 1U;
+	timing.propSeg = data->timing.prop_seg - 1U;
+
+	timing.fpreDivider = data->timing_data.prescaler - 1U;
+	timing.frJumpwidth = data->timing_data.sjw - 1U;
+	timing.fphaseSeg1 = data->timing_data.phase_seg1 - 1U;
+	timing.fphaseSeg2 = data->timing_data.phase_seg2 - 1U;
+	timing.fpropSeg = data->timing_data.prop_seg - 1U;
+	FLEXCAN_SetFDTimingConfig(config->base, &timing);
+
+	data->started = true;
+
+	return 0;
+}
+
+static int mcux_flexcan_fd_stop(const struct device *dev)
+{
+	const struct mcux_flexcan_fd_config *config = dev->config;
+	struct mcux_flexcan_fd_data *data = dev->data;
+	can_tx_callback_t function;
+	void *arg;
+	int alloc;
+	int err;
+
+	if (!data->started) {
+		return -EALREADY;
+	}
+
+	data->started = false;
+
+	/* Abort any pending TX frames before entering freeze mode */
+	for (alloc = 0; alloc < MCUX_FLEXCAN_MAX_TX; alloc++) {
+		function = data->tx_cbs[alloc].function;
+		arg = data->tx_cbs[alloc].arg;
+
+		if (atomic_test_and_clear_bit(data->tx_allocs, alloc)) {
+			FLEXCAN_TransferFDAbortSend(config->base, &data->handle,
+						    ALLOC_IDX_TO_TXMB_IDX(alloc));
+
+			function(dev, -ENETDOWN, arg);
+			k_sem_give(&data->tx_allocs_sem);
+		}
+	}
+
+	FLEXCAN_EnterFreezeMode(config->base);
+
+	if (config->phy != NULL) {
+		err = can_transceiver_disable(config->phy);
+		if (err != 0) {
+			LOG_ERR("failed to disable CAN transceiver (err %d)", err);
+			return err;
+		}
+	}
+
+	return 0;
+}
+
+static int mcux_flexcan_fd_set_mode(const struct device *dev, can_mode_t mode)
+{
+	const struct mcux_flexcan_fd_config *config = dev->config;
+	struct mcux_flexcan_fd_data *data = dev->data;
+	uint32_t ctrl1;
+	uint32_t mcr;
+	uint32_t fdctrl;
+
+	if (data->started) {
+		return -EBUSY;
+	}
+
+
+	if ((mode &
+	     ~(CAN_MODE_FD | CAN_MODE_LOOPBACK | CAN_MODE_LISTENONLY | CAN_MODE_3_SAMPLES)) != 0) {
```

Comment:
Where do you handle the `CAN_MODE_FD` flag? To me it looks like this driver does not handle switching between classic and CAN-FD mode?

Reply:
Fixed this issue now. [Here](https://github.com/dipakgmx/zephyr/blob/feature_mimxrt1060_can_fd_support/drivers/can/can_mcux_flexcan_fd.c#L238) I allow transmission of CAN 2.0 frames.

Reply:
Fixed this now. Am able to use the driver in classic CAN mode.